[Rust] add enums concept exercise

efx opened the pull request:
closes #830

efx committed:
[Rust] add enums concept exercise

closes #830

efx commented on languages/rust/exercises/concept/enums/.meta/example.rs:
```diff
@@ -0,0 +1,28 @@
+    }
+}
+pub fn output_for_short_log(kind: LogLevel, log: &str) -> String {
+    let code = match kind {
```
NOTE: We should be able to declare the integer representation on the `enum` definition?

coriolinus replied:
Define the enum with `#[repr(u8)]`; that should let you do `let code = kind as u8`.

efx commented on languages/rust/exercises/concept/enums/.meta/example.rs:
```diff
@@ -0,0 +1,28 @@
+}
+
+pub fn parse_log_level(log: &str) -> LogLevel {
+    if log.starts_with("[ERROR]") {
```
NOTE: up for a better way to solve this. This is somewhat repetitive.

efx commented on languages/rust/exercises/concept/enums/.docs/hints.md:
```diff
@@ -0,0 +1,21 @@
+
+
+### 1. Parse log level
+
```
I'll submit a follow up PR to complete these sections.

efx commented:
I'm not sure why the link checker fails for https://doc.rust-lang.org/book/ch06-01-defining-an-enum.html. It says a 400 status code despite that URL being accessible.
https://github.com/exercism/v3/pull/883/checks?check_run_id=494807999#step:6:25

coriolinus commented on languages/rust/exercises/concept/enums/.docs/instructions.md:
```diff
@@ -0,0 +1,54 @@
+- `Warning`
+- `Error`
+
+Next, write a function to parse the log level of a log line:
```
By far the simplest approach to this particular problem uses regular expressions. That's not a problem when students are solving this exercise on their own machines; they can just import the regex crate. However, that doesn't work when  solving exercises online: the online editor doesn't support editing `Cargo.toml` or downloading additional crates. 

Perhaps we should invert the sense of this exercise so students are emitting structured log messages, instead of parsing text log lines, to reduce that problem?

efx replied:
> By far the simplest approach to this particular problem uses regular expressions. 

I agree. `&str` methods work but feel verbose.

> Perhaps we should invert the sense of this exercise so students are emitting structured log messages, instead of parsing text log lines, to reduce that problem?

Good idea! Would this be emitting a struct with a `type` field set to the `enum`? Or could you post a snippet demonstrating what you were thinking?

Given #663, it seems like we could simplify as well to fit Rust's idioms better.

I am unclear on how far we should deviate from the csharp concept exercises; I'm leaning toward copying them more verbosely at first to quickly finish a few so we can refine the concepts further and write the concept exercise implementation guide.

coriolinus replied:
Well, we can't do real structured logs if the students haven't done at least declarative macros, so let's call these semistructured. The idea is that we provide the `LogLevel` enum (omitting `Unknown`) and then require students to implement methods like this:

```rust
pub fn log(level: LogLevel, message: &str) -> String { unimplemented!() }
pub fn info(message: &str) -> String { unimplemented!() }
pub fn warn(message: &str) -> String { unimplemented!() }
pub fn error(message: &str) -> String { unimplemented!() }
```

TBH, I'm not a huge fan of the short-log portion of this exercise: while it teaches that some enums can be represented by integer primitives, actually converting between enums and integers in Rust is an antipattern in every case except serialization. I'd just skip that.

coriolinus commented on languages/rust/exercises/concept/enums/src/lib.rs:
```diff
@@ -0,0 +1,9 @@
+pub fn parse_log_level(log: &str) -> LogLevel {
+    unimplemented!("parse a log and return the level")
+}
+pub fn output_for_short_log(kind: LogLevel, log: &str) -> &str {
```
```suggestion
pub fn output_for_short_log(kind: LogLevel, log: &str) -> String {
```

coriolinus commented on languages/rust/exercises/concept/enums/src/lib.rs:
```diff
@@ -0,0 +1,9 @@
+// TODO define enum variants
+pub enum LogLevel {}
+
+pub fn parse_log_level(log: &str) -> LogLevel {
```
```suggestion
pub fn parse_log_level(log: &str) -> Option<LogLevel> {
```

We should demonstrate correct practice for the case that the input string doesn't match the regex `^\[(?P<level>[^\]]*)\]:\s*(?P<message>.*)$`.

The difference between `None` and `Some(Unknown(_))`:

`[foo]: a message foo` -> `Some(Unknown("foo"))`
`not a structured log line` -> `None`

iHiD replied:
So if this is added, we'll need a prereq adding for Option types and an exercise around that, or this exercise to teach Option Types and have that in the different README files

coriolinus replied:
That's a circular dependency: Option type are enums.

iHiD replied:
OK, cool. So this exercise should probably teach Option types then, and have it added to what it teaches etc :)

coriolinus commented on languages/rust/exercises/concept/enums/src/lib.rs:
```diff
@@ -0,0 +1,9 @@
+// TODO define enum variants
+pub enum LogLevel {}
```
Given that we explicitly refer to each variant in the test cases, it's only fair to define them here. The student won't get much from wading through compiler errors to figure out what they should write here that they wouldn't get from reading a properly predefined enum.

iHiD replied:
If they don't write the enum, then (if I'm reading the example.rs correctly), the only code they write to solve the exercise is an if/else. Based on the idea that learning-by-doing is the most effective way to learn, I think that we want someone to be learning how to write the actual enum (and struggling through that if they do), rather than them reading it and hoping they remember it.

coriolinus commented on languages/rust/exercises/concept/enums/tests/enums.rs:
```diff
@@ -0,0 +1,63 @@
+#[test]
+#[ignore]
+fn parse_unknown() {
+    assert_eq!(LogLevel::Unknown, parse_log_level("[FATAL]: Crash!"));
```
```suggestion
    assert_eq!(LogLevel::Unknown("FATAL".into()), parse_log_level("[FATAL]: Crash!"));
```

I don't think it's appropriate to just discard the intended log level in the unknown case.

efx committed:
Update languages/rust/exercises/concept/enums/src/lib.rs

Co-Authored-By: Peter Goodspeed-Niklaus <coriolinus@users.noreply.github.com>

ErikSchierboom commented:
FYI We're in the progress of tweaking the C# Enums exercise a bit: https://github.com/exercism/v3/pull/864#issuecomment-597045987

efx commented:
Thanks for the reviews and discussion folks. I've been slammed with product work and personal matters this week so am going to hopefully address the bulk of this over the weekend.
The key things I'm going to do is:
- simplify *how* we present the concept of enums taking from the C# revision and @coriolinus 's feedback

efx committed:
Update languages/rust/exercises/concept/enums/tests/enums.rs

Co-Authored-By: Peter Goodspeed-Niklaus <coriolinus@users.noreply.github.com>

efx committed:
Update languages/rust/exercises/concept/enums/src/lib.rs

Co-Authored-By: Peter Goodspeed-Niklaus <coriolinus@users.noreply.github.com>

efx committed:
Merge branch 'master' into close-830

efx committed:
Merge branch 'master' into close-830

efx committed:
Merge branch 'master' into close-830

efx committed:
[Rust] invert the sense of enums concept exercise

coriolinus commented on languages/rust/exercises/concept/enums/.docs/after.md:
```diff
@@ -0,0 +1,4 @@
+Rust enums are also called algebraic data types. Why does this matter? Enumerations are a popular concept and programming languages implement them differently. Functional programming has influenced Rust's design and the `enum` implementation demonstrates that.
```
```suggestion
Rust enums are also called [algebraic data types](https://en.wikipedia.org/wiki/Algebraic_data_type), or [tagged unions](https://en.wikipedia.org/wiki/Tagged_union). Why does this matter? Enumerations are a popular concept and programming languages implement them differently. Functional programming has influenced Rust's design and the `enum` implementation demonstrates that.
```

Add more detail.

coriolinus commented on languages/rust/exercises/concept/enums/.meta/design.md:
```diff
@@ -0,0 +1,44 @@
+
+## Out of scope
+
+- Tuple enums.
```
```suggestion
- Tuple and struct enums.
```

coriolinus commented on languages/rust/exercises/concept/enums/.meta/design.md:
```diff
@@ -0,0 +1,44 @@
+This exercise's prerequisites Concepts are:
+
+- `strings-basic`: log lines are `string` values.
+- `numbers-basic`: enum values are assigned a numeric value.
```
The fact that an enum discriminates between its variants by assigning each a numeric tag is an implementation detail, and should not be part of the basic introduction to enums.

```suggestion
```

coriolinus commented on languages/rust/exercises/concept/enums/.meta/design.md:
```diff
@@ -0,0 +1,44 @@
+
+- `strings-basic`: log lines are `string` values.
+- `numbers-basic`: enum values are assigned a numeric value.
+- `conditionals-if`: used to handle the case when the enum could not be parsed.
```
We are no longer parsing or using `if` conditionals in this exercise.

```suggestion
```

coriolinus reviewed (approved):
Looks good! A few quibbles below.

One more thing I think students would find handy: a feature-gated test intended to teach them how to add variants to enums, and that the Rust compiler has their back in that case. Something like this:

**README.md**:
```markdown
There is a feature-gated test in this suite. Feature gates disable compilation entirely for certain sections of your program. They will be covered later, but for now, just know that there is a test which is only built and run when you use a special testing invocation:

    cargo test --features add-a-variant

This test is meant to show you how to add a variant to your enum.
```

**Cargo.toml**:
```toml
[features]
add-a-variant = []
```

**tests/enums.rs**:
```rust
#[test]
#[cfg(feature="add-a-variant")]
fn add_a_variant() {
    // this test won't even compile until the enum is complete, which is why it is feature-gated.
    assert_eq!(
        log(LogLevel::Debug, "reached line 123"),
        "[DEBUG]: reached line 123",
    );
}
```

efx committed:
Update languages/rust/exercises/concept/enums/.docs/after.md

Co-Authored-By: Peter Goodspeed-Niklaus <coriolinus@users.noreply.github.com>

efx committed:
Update languages/rust/exercises/concept/enums/.meta/design.md

Co-Authored-By: Peter Goodspeed-Niklaus <coriolinus@users.noreply.github.com>

efx committed:
Update languages/rust/exercises/concept/enums/.meta/design.md

Co-Authored-By: Peter Goodspeed-Niklaus <coriolinus@users.noreply.github.com>

efx committed:
Update languages/rust/exercises/concept/enums/.meta/design.md

Co-Authored-By: Peter Goodspeed-Niklaus <coriolinus@users.noreply.github.com>

efx commented:
@coriolinus grand, thank you for the updated review! I think your suggestion for the feature gated variant directly addresses the concerns @iHiD raised earlier about students learning by doing. I'll add that in!

efx committed:
add description for other variant

iHiD commented:
How will feature-gates work in v3 in the in-browser coding?

coriolinus commented:
They would need support from the front end. It's not difficult to adjust
the test runner to pass through build arguments like that, but the front
end would need to have a checkbox or something (maybe defined in a
configuration file for the exercise?) with which the student can set the
build argument.

Alternately, the front end could let the student set the testing command
string entirely, which makes config a lot easier and helps get the student
more comfortable on the command line.

On Mon, Mar 30, 2020, 18:53 Jeremy Walker <notifications@github.com> wrote:

> How will feature-gates work in v3 in the in-browser coding?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/v3/pull/883#issuecomment-606117026>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB3V4TQZ5ETOJR2WZIYOMCTRKDFBHANCNFSM4LEFRT6A>
> .
>

efx commented:
Where is the front end code / is it publicly available?

workingjubilee commented:
You can see it here: https://research.exercism.io
And at least most of the code: https://github.com/exercism/research_experiment_1

ErikSchierboom commented:
> And at least most of the code: https://github.com/exercism/research_experiment_1

This is actually not the source for the front end code, but just for the exercises used in the research experiment. The actual source code for the website can be found here: https://github.com/exercism/website
